<a href="https://colab.research.google.com/github/divosiber/Project-NLTK-Bag-Of-Words/blob/master/Text_Classification_Bag_of_words.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Implementation of Artificial Intelligence

Text classification berdasarkan Bag-of-Words

- tiga topik terkait: Kesehatan, Olahraga, Keuangan.

Klasifikasikan teks ini dalam topik :
“Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres.”

# **Mengimport dan mendownload library nltk**

In [ ]:
import nltk
nltk.download('popular')

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cmudict.zip.
[nltk_data]    | Downloading package gazetteers to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gazetteers.zip.
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/genesis.zip.
[nltk_data]    | Downloading package gutenberg to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gutenberg.zip.
[nltk_data]    | Downloading package inaugural to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/inaugural.zip.
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/movie_reviews.zip.
[nltk_data]    | Downloading package names to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/names.zip.
[nltk_data]    | Downloading package shakespeare to /root/nltk_data...
[nlt

True

# **import library**

Import Library Tokenize untuk memisahkan teks menjadi beberwpa bagian.

import stopwords untuk menghapus beebrapa kata yang tidak akan digunakan


import library limitization

In [ ]:
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

# **Sumber artikel**

Source article untuk  artikel sports = https://theathletic.com/5435913/2024/04/23/erik-ten-hag-thinks-manchester-united-are-unlucky-hes-only-partly-right/?source=nyt_sports

sumber artikel finance = https://hbr.org/2014/09/how-to-prioritize-your-innovation-budget

sumber artikel medical = https://www.path.org/our-impact/articles/how-does-human-milk-banking-help-ensure-child-health/




# **Klasifikasi teks artikel dengan Bag of words**

# **Data Latih**

In [ ]:
#data latih
olahraga = 'Ten Hag took much the same line of argument before United’s last Premier League outing against Bournemouth. While accepting that “like a minister” he will bear ultimate responsibility for results, he could not help but bemoan his side’s bad luck over the past eight months.“It’s huge. A lot went against us this season,” he said. And though United’s misfortune is not limited to refereeing calls in Ten Hag’s mind, that was where he trained his focus.“You see all the penalties we conceded last week (against Chelsea and Liverpool) could also have been going in another way. You think over the course of a season sometimes you will get one, sometimes you will concede one. This season it feels like we only concede.United have been awarded five penalties this season and have conceded 11, with four given away in the opening four games of the Champions League group stage. While most of those in Europe were not especially contentious, many of the six conceded in the Premier League have sparked debate.Some have been soft — Rasmus Hojlund and Casemiro’s concessions against Manchester City and Wolverhampton Wanderers in particular — and others more debatable. None, it should be noted, have resulted in the officials responsible being stood down for the subsequent round of fixtures, as happened after Wolves were denied a penalty at Old Trafford on the opening weekend of the season.All those decisions, however, are a matter of opinion. Outside of offside, most refereeing calls are subjective by nature and, as the era of VAR has taught us, there are different definitions of what constitutes a clear and obvious mistake.Ten Hag has more substantive grounds for complaint on arguably the biggest single reason for United’s struggles: player injuries and enforced absences. The revolving door of United’s treatment room has seen all but four senior squad members — Bruno Fernandes, Andre Onana, Diogo Dalot and Alejandro Garnacho — pass through it at some point this year.The 2-2 draw at Bournemouth was the first time United have named an unchanged line-up since the opening two games of a season ravaged by injury. According to data from transfermarkt, United’s squad have collectively spent 1,710 days sidelined since the start of the season.Ten Hag said last week he has not been able to pick his “favourite” line-up since the 2-1 victory over Manchester City at Old Trafford in January of last year. Just as United’s injuries have appeared to abate, new concerns have cropped up.Fresh problems for Willy Kambwala, Mason Mount and Sofyan Amrabat meant United’s absentee list swelled into double figures again ahead of the semi-final, while Marcus Rashford and Scott McTominay both appeared to be carrying issues when substituted at Wembley.The absence of either of his first-choice left-backs for the majority of the season has, Ten Hag feels, had a material effect on United’s ability to play the way he wants. Lisandro Martinez’s unavailability has deprived him of a player who had a transformative effect during his first year in Manchester.But is it all down to luck or could certain things be done differently? United have set to work restructuring the medical department since the appointment of head of sports medicine Gary O’Driscoll. Sources, who asked to remain anonymous to protect their relationships, believe there have been noticeable improvements since the former Arsenal club doctor’s arrival and that restructuring continues apace.Ten Hag’s training methods have also come under scrutiny and can be intense, particularly for those not involved in matches, who are put through rigorous sessions the day after games to maintain a consistent level of physical load across the squad. The fast, direct and often chaotic style of play that has been adopted this season also has to be considered as part of that equation.And when looking at the performances that style of play has produced on a statistical level, United have arguably had their fair share of good fortune to balance out the bad.Everybody knows by now that United face a lot of shots on goal — 574 in total in the Premier League this season. No top-flight team has faced as many on a per-game basis, but in the context of recent history, that figure only becomes all the more remarkable.Since 2016-17, eight of the 15 top-flight sides to have faced more shots than United have been relegated. None have finished higher than 15th. At the current rate, United will surpass all of those 15 sides and yet even in the absolute worst-case scenario, they cannot finish any lower than 14th.Ten Hag has defended United’s apparent willingness to give up shots by arguing they are predominantly low-quality chances and he has a point. The average shot United have faced in the league this season has had a 10 per cent chance of resulting in a goal.Brentford and Newcastle have the worst record in that regard, with the average shot having a 13 per cent chance of being scored. The difference between a 13 per cent and 10 per cent chance is small but significant. A marginal gain, if you like.But if you concede at least 20 shots a game, as United have regularly been doing of late, and one in every 10 goes in, you’ll need to score three to win. The eighth-worst attack in the league, with only 47 goals in 32 league games, cannot count on that.United’s 47 goals is level with Luton Town and in line with expected data, too. Defensively, Ten Hag’s side have conceded 48 goals — one of the Premier League’s better records — but from an expected total of 59.8.Take one away from the other and United’s expected goal difference is -12.2, the fifth-worst in the league. Suddenly, that actual goal difference of -1 does not look so bad after all.But nothing can change perceptions and narratives around a side like a favourable run of fixtures, in the short term at least, and United now face the Premier League’s bottom two at Old Trafford in the space of four days.It should not need saying, but United are a better side than both Sheffield United and Burnley by any comprehensive measure. They should not need to get lucky to prove it.'
keuangan = 'Here’s the scene: A problem has come up with one of your supply chain vendors, threatening to delay timely shipment of your product. At the same time, a potential opportunity appears that, with some exploration and investment, could lead to a new generation of products down the road. Which do you respond to first?You probably reach for your firefighter’s hat to extinguish the short-term problem. And therein lies a bigger problem. Leaders and organizations are under more stress than ever to do two things simultaneously: deliver on today’s pressing commitments by troubleshooting and refining processes; and find and invest in innovation opportunities that will create tomorrow’s success. How your organization responds to this stress in allocating scarce resources is a crucial but often unaddressed issue. The natural bias is to respond immediately to what is in front of you (like answering endless emails as they come in, for instance). The problem is, this instinct crowds out longer term, innovative thinking.We’ve talked to many organizations in this bind. One of them told us, “We’re playing non-stop ‘Whac-A-Mole’ here.” At another, the unfortunate mantra was, “The urgent drives out the important.” But we’ve found that many leading organizations are able to overcome this bias, diverting significant resources away from today’s requirements to fund the innovations that will deliver tomorrow’s value. To find out how they do this, we focused on the two key questions underlying the challenge: How much is your organization spending on innovation? And how much do you think it should be spending?This seems basic, but often people just don’t know the precise answer or haven’t thought in these simple terms. When we recently put these questions to a CEO, he said he guessed his organization was currently allocating 5% of their spending to innovation (new products and services), but added that they really should be investing 10-15%. He went on to say that the insatiable demands of today’s operational turbulence were robbing him and his organization of ability to invest in the future.Segregate funds for improvement and innovation. You need to measure spending across the four categories, and then be disciplined about segregating funds for improvement and innovation. In the absence of this clear segregation, the turbulence of day-to-day operations will devour the lion’s share of resources. Go ahead and do the math on your current allocation; you’ll likely find that more than 90% of your resources are devoted to keeping the lights on. Many leading IT organizations have recognized this problem and manage their budgets in three buckets: operations, maintenance, and innovation. They aggressively attempt to drive down the portion of their spending on operations and maintenance from 90% to 60%.Create new organizations and controls for innovation. You need to develop organizational structures and controls that are appropriate for both sustaining and disruptive innovations. Today’s operations and incremental improvements can be managed within the traditional management structure, a command-and-control hierarchy, but tomorrow’s sustaining innovations and big, disruptive innovations need new organizational structures and controls. It’s too easy for revenue-producing parts of the business to poach resources from innovation projects and teams that are not (yet) contributing to the top line. They need to be protected – made autonomous, with their own dedicated budgets, resources, and leadership – until they are. And innovations need measures and controls which reflect their experimental approach to learning as they chart new territory with unpredictable outcomes. Performance should be measured by market momentum (such as new targeted customers or partners, deal size, and PR buzz).In the battle between today and tomorrow, today will win every time unless the organization consciously, strategically decides to extend a helping hand to tomorrow. The first step is to measure spending against the four choices above. Do you like what you see? Are you willing to do something about it if you don’t? Remember, you’re making a crucial choice about your company’s future.'
kesehatan = 'Superior to formula, human milk provides essential nutrients for infant growth and development. According to UNICEF, “feeding infants nothing but breastmilk guarantees them a food source that is uniquely adapted to their nutrient needs, while also being safe, clean, and healthy.” Despite the nutritional benefits of human milk, less than half of infants under six months old are exclusively breastfed.Infant mortality remains high globally as millions of babies die within their first year of life. Unfortunately, numerous social, political, and structural barriers can hinder a mother’s ability to breastfeed. These often include the influence of social norms, lack of support for breastfeeding mothers, and impact of infant formula marketing. Additionally, women may lack knowledge about the benefits of breastfeeding or are simply unable to produce enough breastmilk. To help address some of these barriers, PATH is working collaboratively with governments and partners worldwide to increase equitable access to human milk.Human milk banks play an essential role in newborn health by providing human donor milk to newborns who would otherwise not be able to receive their mothers’ own milk. This is necessary for newborns that are small and/or sick, meaning those that are born preterm, are small for their gestational age, have low birth weight, have an illness or birth complication, as well as those requiring hospitalization. Donor human milk is specifically targeted to this population of newborns.Donor human milk is used as a bridge, while the mother’s lactation supply can be established and the infant transitions to feeding directly from the breast, equipping both mother and baby for healthy breastfeeding post-discharge. Milk banks collect, screen, store, process, and distribute human milk. Women who donate milk usually breastfeed their own infants and have a milk supply that exceeds their own infants’ needs. Best practice is to implement human milk banks as a component within a comprehensive and integrated system, to ensure they are in line with the broader purpose of protecting, promoting and supporting breastfeeding and child nutrition.Human milk banks are increasing worldwide, as the global health community recognizes the benefits of such efforts on newborn health. PATH works at a global level advocating for exclusive human milk diets and with Ministries of Health to ensure a sustainable and locally appropriate human milk bank system is embedded within newborn care and breastfeeding promotion—such as our work in Kenya. Furthermore, PATH has resources available for establishing and integrating human milk banks into existing health facilities to expand global access. PATH toolkits detail how to ensure appropriate use of donor human milk and how to introduce a rigorous human milk bank system for screening, pasteurizing, storage and distribution, all of which is required to ensure safe donor human milk. These essential efforts help to provide equitable access to human milk for babies around the world.Given what we know about the nutritional benefits of human milk, and the accessibility challenges of this feeding option for newborns, exclusive human milk diets contribute to preventing and reducing diarrheal disease. Infants are at greatest risk of diarrheal disease when they are given foods other than breast milk. This is because they are more likely to be exposed to food- and water-borne pathogens and lose the protection of breast milk’s anti-infective properties. Furthermore, diarrhea is a leading cause of malnutrition in children under five, and malnourished children are more vulnerable to infection. Human milk helps an infant’s immune system to safeguard against infectious diseases like diarrhea and other dangerous pathogens.Breastfeeding and safe provision of donor human milk are significant in efforts to defeat diarrheal disease­: we encourage stakeholders, health providers, and other advocates to strengthen systems to support exclusive human milk diets and utilize available milk banking resources, when appropriate, to advocate for expanded access to nutritional breastmilk for infants in all regions of the world.'

# **Text processing**

setelah mendapatkan artikel selanjutnya kami membuat variabel baru untuk menampung variabel di list yang telah dibuat. selanjutnya melakukan looping per artikel untuk  di word teknize dimana fungsinya untuk memotong teks menjadi kata - kata, melakukan lemmatization, dan stopwords untuk menghapus kata kata yang ada di artikel yang tidak penting untuk menghemat memorui.

In [ ]:
lemmatizer = WordNetLemmatizer()

texts = [olahraga, kesehatan, keuangan]
bow_keys = []
corpus_texts = []
for text in texts:
    words  = word_tokenize(text)
    texts = [lemmatizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))]
    bow_keys += texts
    text = ' '.join(texts)
    corpus_texts.append(text)
bow_keys = set(bow_keys)
print(bow_keys)
print(corpus_texts)

{'like', 'crowd', 'collect', 'protecting', 'Sheffield', ')', 'play', 'significant', 'up.Fresh', 'Scott', 'territory', 'PATH', 'Premier', 'sport', 'Lisandro', 'opening', 'set', 'develop', 'Champions', 'Donor', 'growth', 'Additionally', 'equipping', 'term', 'match', 'chance', 'And', 'another', 'preterm', 'safeguard', 'region', 'going', 'believe', 'spent', 'protect', 'fifth-worst', 'age', 'structure', 'share', 'swelled', 'diarrheal', 'unavailability', 'four', 'sustainable', 'Infants', 'argument', 'decision', 'goal.Brentford', 'Remember', 'season.All', 'Milk', 'provision', '—', 'exposed', 'like.But', 'timely', 'structural', 'potential', 'department', 'system', 'lead', 'support', 'unfortunate', 'squad', 'attempt', 'mantra', 'email', 'improvement', 'well', 'Diogo', 'week', 'differently', 'favourite', 'easy', 'newborn', 'United', 'likely', 'produce', 'extend', 'data', 'He', 'birth', 'consistent', 'rate', 'particular', 'style', 'single', 'count', 'available', 'Liverpool', 'people', 'formula', 

# **Teks yang akan di klasifikasikan**

“Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres.”

selanjutnya mengambil data teks untuk di uji apakah teks tersebut termasuk dalam klasifikasi sport,medical, atau finance. pertama teks tersebut di tokenize terlebih dahulu menggunakan library word_tokenize, selanjutnya di limitization, setelah itu menggunkan stopword untuk menghapus kata yang tidak penting.Dan terakhir membat vektor dari library countvectorizer yang ada di sklearn

In [ ]:
text_clf = 'Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres'
wt = word_tokenize(text_clf)
word_lematizer = [lemmatizer.lemmatize(word) for word in wt if word not in set(stopwords.words('english'))]
words_corpus = [word for word in word_lematizer if word in set(bow_keys)]
qtc = ' '.join(words_corpus)
corpus_texts.append(qtc)

from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
bow_vectors = cv.fit_transform(corpus_texts).toarray()
print(bow_vectors)
print(len(bow_vectors[0]))

[[3 1 1 ... 3 1 2]
 [0 0 0 ... 1 0 0]
 [1 0 0 ... 0 1 3]
 [0 0 0 ... 0 0 0]]
817


**seperti terlihat pada vektor di atas, total kata yang digunakan adalah 817 kata**





Step selanjutnya adalah normalisasi Bag of Words, setelah di noemalisasi selanjutnya mengecek kesamaan menggunakan dot product atau perkalian vektor. dan melihat kesamaan teks yang telah di uji tersebut termasuk dalam klasifikasi apa

In [ ]:
bow_texts_norm = []
for bow in bow_vectors:
  length = (sum(i*i for i in bow)) ** 0.5
  bow_norm = bow / length
  bow_texts_norm.append(bow_norm)

similarity_vector = []
bow_norm_query = bow_texts_norm[3]
for bow in bow_texts_norm[:3]:
  similarity_vector.append(sum(i*j for i,j in zip(bow,bow_norm_query)))
print(similarity_vector)

id_max_sim = similarity_vector.index(max(similarity_vector))

if (id_max_sim == 1):
  print ("Teks diklasifikasikan sebagai: Medical (Kesehatan)")
elif (id_max_sim == 0):
  print ("Teks diklasifikasikan sebagai : Sport (olahraga)")
elif (id_max_sim == 2):
  print ("Teks diklasifikasikan sebagai: Finance (keuangan)")

[0.41801910277201215, 0.004782554451485865, 0.021213203435596423]
Teks diklasifikasikan sebagai : Sport (olahraga)


# **kesimpulan**

setelah melakukan perulagan bercabang, Teks Classification menggunakan Bag Of Words dengan data uji tersebut termasuk dalam kategori klasifikasi sport atau olahraga